epochStart/metachain/auctionListSelector.go

```diff
@@ -0,0 +1,255 @@
+		"auction list size", auctionListSize,
+		fmt.Sprintf("available slots (%v -%v)", maxNumNodes, numOfValidatorsAfterShuffling), availableSlots,
+	)
+
+	err = als.sortAuctionList(auctionList, randomness)
+	if err != nil {
+		return err
+	}
+
+	numOfAvailableNodeSlots := core.MinUint32(auctionListSize, availableSlots)
+	als.displayAuctionList(auctionList, numOfAvailableNodeSlots)
+
+	for i := uint32(0); i < numOfAvailableNodeSlots; i++ {
+		newNode := auctionList[i]
+		newNode.SetList(string(common.SelectedFromAuctionList))
+		err = validatorsInfoMap.Replace(auctionList[i], newNode)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// TODO: Move this in elrond-go-core
+func safeSub(a, b uint32) (uint32, error) {
+	if a < b {
+		return 0, core.ErrSubtractionOverflow
+	}
+	return a - b, nil
+}
+
+func getAuctionListAndNumOfValidators(validatorsInfoMap state.ShardValidatorsInfoMapHandler) ([]state.ValidatorInfoHandler, uint32) {
+	auctionList := make([]state.ValidatorInfoHandler, 0)
+	numOfValidators := uint32(0)
+
+	for _, validator := range validatorsInfoMap.GetAllValidatorsInfo() {
+		if validator.GetList() == string(common.AuctionList) {
+			auctionList = append(auctionList, validator)
+			continue
+		}
+		if isValidator(validator) {
+			numOfValidators++
+		}
+	}
+
+	return auctionList, numOfValidators
+}
+
+func (als *auctionListSelector) sortAuctionList(auctionList []state.ValidatorInfoHandler, randomness []byte) error {
+	if len(auctionList) == 0 {
+		return nil
+	}
+
+	validatorTopUpMap, err := als.getValidatorTopUpMap(auctionList)
+	if err != nil {
+		return fmt.Errorf("%w: %v", epochStart.ErrSortAuctionList, err)
+	}
+
+	pubKeyLen := len(auctionList[0].GetPublicKey())
+	normRandomness := calcNormRand(randomness, pubKeyLen)
+	sort.SliceStable(auctionList, func(i, j int) bool {
+		pubKey1 := auctionList[i].GetPublicKey()
+		pubKey2 := auctionList[j].GetPublicKey()
+
+		nodeTopUpPubKey1 := validatorTopUpMap[string(pubKey1)]
+		nodeTopUpPubKey2 := validatorTopUpMap[string(pubKey2)]
+
+		if nodeTopUpPubKey1.Cmp(nodeTopUpPubKey2) == 0 {
+			return compareByXORWithRandomness(pubKey1, pubKey2, normRandomness)
+		}
+
+		return nodeTopUpPubKey1.Cmp(nodeTopUpPubKey2) > 0
+	})
+
+	return nil
+}
+
+func (als *auctionListSelector) getValidatorTopUpMap(validators []state.ValidatorInfoHandler) (map[string]*big.Int, error) {
+	ret := make(map[string]*big.Int, len(validators))
+
+	for _, validator := range validators {
+		pubKey := validator.GetPublicKey()
+		topUp, err := als.stakingDataProvider.GetNodeStakedTopUp(pubKey)
+		if err != nil {
+			return nil, fmt.Errorf("%w when trying to get top up per node for %s", err, hex.EncodeToString(pubKey))
+		}
+
+		ret[string(pubKey)] = topUp
+	}
+
+	return ret, nil
+}
+
+func calcNormRand(randomness []byte, expectedLen int) []byte {
+	rand := randomness
+	randLen := len(rand)
+
+	if expectedLen > randLen {
+		repeatedCt := expectedLen/randLen + 1
```

Comment:
might add a comment or at least unit tests. Here are some of them to get you started:
```GO
func TestCalcNormRand(t *testing.T) {
	t.Parallel()

	t.Run("randomness longer than expected len", func(t *testing.T) {
		t.Parallel()

		randomness := []byte("rand")
		expectedLen := 2

		result := calcNormRand(randomness, expectedLen)

		require.Equal(t, []byte("ra"), result)
	})

	t.Run("randomness length equal to expected len", func(t *testing.T) {
		t.Parallel()

		randomness := []byte("rand")
		expectedLen := 4

		result := calcNormRand(randomness, expectedLen)

		require.Equal(t, []byte("rand"), result)
	})

	t.Run("randomness length less than expected len", func(t *testing.T) {
		t.Parallel()

		randomness := []byte("rand")
		expectedLen := 6

		result := calcNormRand(randomness, expectedLen)

		require.Equal(t, []byte("randra"), result)
	})
}
```

Reply:
Good suggestion; Added the tests